File: localities.py
# Populate location database tables - run stance_grouping.py first to generate localities.json

import json
from typing import TypedDict, Dict
from xml.etree.ElementTree import *

import pandas as pd
from defusedxml.ElementTree import parse
import sqlite3


class StopData(TypedDict):
    ATCOCode: str
    Lat: float
    Long: float
    Street: str
    Indicator: str


# Parse locality XML into tuples
def convert_locality(locality):
    code = locality.findtext(".//{http://www.naptan.org.uk/}NptgLocalityCode")
    name = locality.findtext(".//{http://www.naptan.org.uk/}LocalityName")
    qualifier = locality.findtext(".//{http://www.naptan.org.uk/}QualifierName")
    parent = locality.findtext(".//{http://www.naptan.org.uk/}ParentNptgLocalityRef")
    long = locality.findtext(".//{http://www.naptan.org.uk/}Longitude")
    lat = locality.findtext(".//{http://www.naptan.org.uk/}Latitude")
    if long is not None:
        long = float(long)
    if lat is not None:
        lat = float(lat)
    return code, name, qualifier, parent, long, lat


def get_crs_codes() -> Dict[str, str]:
    df: pd.Series = pd.read_csv("crs.csv", index_col="ATCOCode")["CrsRef"]
    return df.to_dict()


def init_db(db: sqlite3.Connection):
    print("Initialise database")
    # Initialise database if it does not exist
    with open("bus-site/gtfs/model.sql") as file:
        script = " ".join(file.readlines())
        db.executescript(script)


def insert_localities(db: sqlite3.Connection):
    print("Insert localities into database")

    # Import NPTG locality data and parse XML
    tree: ElementTree = parse("NPTG.xml")
    root = tree.getroot()
    localities = root.findall(".//{http://www.naptan.org.uk/}NptgLocality")

    # Parse localities into objects
    data = [convert_locality(locality) for locality in localities]

    # Add localities to the database
    db.executemany("REPLACE INTO localities (code, name, qualifier, parent, lat, long) VALUES (?, ?, ?, ?, ?, ?)", data)
    db.execute("DROP TABLE IF EXISTS localities_search")
    db.execute("CREATE VIRTUAL TABLE localities_search USING fts5(name, qualifier, code UNINDEXED)")
    db.execute("INSERT INTO localities_search(name, qualifier, code) SELECT name, qualifier, code FROM localities")


def insert_stops(db: sqlite3.Connection):
    print("Load CSV codes")
    crs_codes = get_crs_codes()

    print("Importing stops")

    # Open the stop data generated by stance_grouping.py
    with open("localities.json", "r") as stops_file:
        stops_data: dict[str, dict[str, list[StopData]]] = json.load(stops_file)

    # Clear existing data
    db.execute("DELETE FROM stops")
    db.execute("DELETE FROM stances")

    # Import stops
    for locality, stops in stops_data.items():
        for stop, stances in stops.items():
            stop_id = db.execute("INSERT INTO stops (name, locality) VALUES (?, ?) RETURNING id", (stop, locality)).fetchone()[0]
            for stance in stances:
                code = stance["ATCOCode"]
                lat = stance["Lat"]
                long = stance["Long"]
                street = stance["Street"]
                indicator = stance["Indicator"]
                crs = crs_codes.get(code)
                db.execute("INSERT INTO stances (code, street, indicator, lat, long, stop, crs) VALUES (?, ?, ?, ?, ?, ?, ?)", (code, street, indicator, lat, long, stop_id, crs))

    print("Generating locality names")

    # Generate locality names based on locality hierarchy
    db.execute("""UPDATE stops SET locality_name =
                     (SELECT GROUP_CONCAT(name, ' › ') FROM (
                        WITH RECURSIVE
                            find_parent_names(level, code) AS (
                                VALUES(0, stops.locality)
                                UNION
                                SELECT level+1, parent FROM localities, find_parent_names
                                WHERE localities.code=find_parent_names.code
                            )
                        SELECT name FROM localities, find_parent_names
                        WHERE localities.code = find_parent_names.code
                        ORDER BY level desc
                    ));
    """)

    print("Setting up stop search")

    # Reset and populate the stop search database
    db.execute("DROP TABLE IF EXISTS stops_search")
    db.execute("CREATE VIRTUAL TABLE stops_search USING fts5(name, parent, qualifier, id UNINDEXED)")
    db.execute("INSERT INTO stops_search(name, parent, qualifier, id) SELECT stops.name, stops.locality_name, qualifier, stops.id FROM stops INNER JOIN localities l on l.code = stops.locality")


def main():
    # Connect to stops database
    db = sqlite3.connect("bus-site/stops.sqlite")

    print("Initialise database")
    # Initialise database if it does not exist
    with open("bus-site/gtfs/model.sql") as file:
        script = " ".join(file.readlines())
        db.executescript(script)

    insert_localities(db)
    insert_stops(db)

    print("Committing to database")
    # Commit to the database
    db.commit()
    db.close()


if __name__ == "__main__":
    main()

File: server/src/bin/ingester/localities.rs
use std::collections::HashMap;
use std::error::Error;
use std::fs;
use std::str::FromStr;
use rusqlite::{Connection, params};
use serde::{Deserialize, Serialize};
use spex::parsing::XmlReader;
use spex::xml::Element;
use BusBoardsServer::download_if_old;

pub type Localities = HashMap<LocalityCode, HashMap<StopName, Vec<Stance>>>;
pub type LocalityCode = String;
pub type StopName = String;

const NAPTAN_NS: &str = "http://www.naptan.org.uk/";

#[derive(Serialize, Deserialize, Clone)]
pub struct Stance {
    #[serde(rename = "ATCOCode")]
    pub atco_code: String,
    #[serde(rename = "Lat")]
    pub lat: f64,
    #[serde(rename = "Long")]
    pub long: f64,
    #[serde(rename = "Street")]
    pub street: Option<String>,
    #[serde(rename = "Indicator")]
    pub indicator: Option<String>,
    #[serde(rename = "Arrival")]
    pub arrival: bool,
    #[serde(rename = "CrsRef")]
    pub crs: Option<String>
}

pub fn load_localities_json() -> Localities {
    let json_str = fs::read_to_string("localities.json").expect("Cannot find localities.json");
    serde_json::from_str(&json_str).expect("JSON parse fail")
}

pub fn insert_localities(db: &mut Connection) -> Result<(), Box<dyn Error>> {
    println!("Insert localities into database");
    let nptg_file = download_if_old("https://naptan.api.dft.gov.uk/v1/nptg", "NPTG.xml")?;
    let xml = XmlReader::parse_auto(nptg_file)?;
    let mut xml_localities = xml.root()
        .pre_ns(NAPTAN_NS)
        .all("NptgLocalities");
    let localities = xml_localities.all("NptgLocality").iter().filter_map(convert_locality);

    db.pragma_update(None, "foreign_keys", "OFF")?;
    let tx = db.transaction()?;
    {
        let mut stmt = tx.prepare("REPLACE INTO localities (code, name, qualifier, parent, long, lat) VALUES (?, ?, ?, ?, ?, ?)")?;
        localities.for_each(|loc| {
            stmt.execute(params![
                loc.code, loc.name, loc.qualifier, loc.parent, loc.long, loc.lat
            ]).expect("Error inserting locality");
        });
    }
    tx.commit()?;
    db.pragma_update(None, "foreign_keys", "ON")?;
    db.execute_batch(r"
        DROP TABLE IF EXISTS localities_search;
        CREATE VIRTUAL TABLE localities_search USING fts5(name, qualifier, code UNINDEXED);
        INSERT INTO localities_search(name, qualifier, code) SELECT name, qualifier, code FROM localities;
    ")?;
    Ok(())
}

fn convert_locality(locality: &Element) -> Option<Locality> {
    Some(Locality {
        code: locality.req(("NptgLocalityCode", NAPTAN_NS)).text().ok().unwrap().to_string(),
        name: locality.pre_ns(NAPTAN_NS).req("Descriptor").req("LocalityName").text().ok().unwrap().to_string(),
        qualifier: locality.pre_ns(NAPTAN_NS).req("Descriptor").req("Qualify").req("QualifierName").text().ok().map(|s| s.to_string()),
        parent: locality.pre_ns(NAPTAN_NS).req("ParentNptgLocalityRef").text().ok().map(|s| s.to_string()),
        long: f64::from_str(locality.pre_ns(NAPTAN_NS).req("Location").req("Translation").req("Longitude").text().ok().unwrap().trim()).ok().unwrap(),
        lat: f64::from_str(locality.pre_ns(NAPTAN_NS).req("Location").req("Translation").req("Latitude").text().ok().unwrap().trim()).ok().unwrap(),
    })
}

pub fn insert_stops(db: &mut Connection) -> Result<(), Box<dyn Error>> {
    println!("Importing stops");
    let localities = load_localities_json();
    db.execute_batch(r"
        DELETE FROM stops;
        DELETE FROM stances;
    ")?;

    let tx = db.transaction()?;
    {
        let mut insert_stop = tx.prepare("INSERT INTO stops (name, locality) VALUES (?, ?) RETURNING id")?;
        let mut insert_stance = tx.prepare("INSERT INTO stances (code, street, indicator, lat, long, stop, crs) VALUES (?, ?, ?, ?, ?, ?, ?)")?;

        for (locality, stops) in localities {
            for (stop, stances) in stops {
                match insert_stop.query_row([&stop, &locality], |row| row.get::<_, u64>(0)) {
                    Ok(stop_id) => {
                        for stance in stances {
                            insert_stance.execute(params![
                                stance.atco_code,
                                stance.street,
                                stance.indicator,
                                stance.lat,
                                stance.long,
                                stop_id,
                                stance.crs
                            ]).expect(format!("Stance insert failed: {}", stance.atco_code).as_str());
                        }
                    },
                    Err(err) => {
                        println!("Stop insert failed: {}, {}, {}", locality, stop, err);
                    }
                }
            }
        }
    }
    tx.commit()?;

    println!("Generating locality names");
    db.execute(r"
        UPDATE stops SET locality_name =
             (SELECT GROUP_CONCAT(name, ' › ') FROM (
                WITH RECURSIVE
                    find_parent_names(level, code) AS (
                        VALUES(0, stops.locality)
                        UNION
                        SELECT level+1, parent FROM localities, find_parent_names
                        WHERE localities.code=find_parent_names.code
                    )
                SELECT name FROM localities, find_parent_names
                WHERE localities.code = find_parent_names.code
                ORDER BY level desc
        ));
    ", [])?;

    /* println!("Setting up stop search");
    db.execute_batch(r"
        DROP TABLE IF EXISTS stops_search;
        CREATE VIRTUAL TABLE stops_search USING fts5(name, parent, qualifier, id UNINDEXED, locality UNINDEXED);
        INSERT INTO stops_search(name, parent, qualifier, id, locality) SELECT stops.name, stops.locality_name, qualifier, stops.id, stops.locality FROM stops INNER JOIN localities l on l.code = stops.locality;
    ")?; */

    Ok(())
}

struct Locality {
    code: String,
    name: String,
    qualifier: Option<String>,
    parent: Option<String>,
    long: f64,
    lat: f64
}